Cabal/Distribution/PackageDescription/Parsec.hs

```diff
@@ -0,0 +1,364 @@
+            parseFailure pos $ "Unknown section: " ++ show name
+            -- parseWarning pos $ "Ignoring section: " ++ show name
+            -- return gpd
+
+{-
+    maybeWarnCabalVersion newsyntax pkg
+      | newsyntax && specVersion pkg < Version [1,2] []
+      = lift $ warning $
+             "A package using section syntax must specify at least\n"
+          ++ "'cabal-version: >= 1.2'."
+
+    maybeWarnCabalVersion newsyntax pkg
+      | not newsyntax && specVersion pkg >= Version [1,2] []
+      = lift $ warning $
+             "A package using 'cabal-version: "
+          ++ displaySpecVersion (specVersionRaw pkg)
+          ++ "' must use section syntax. See the Cabal user guide for details."
+      where
+        displaySpecVersion (Left version)       = display version
+        displaySpecVersion (Right versionRange) =
+          case asVersionIntervals versionRange of
+            [] {- impossible -}           -> display versionRange
+            ((LowerBound version _, _):_) -> display (orLaterVersion version)
+
+    maybeWarnCabalVersion _ _ = return ()
+-}
+
+    handleFutureVersionParseFailure :: Version -> ParseResult a -> ParseResult GenericPackageDescription
+    handleFutureVersionParseFailure cabalVersionNeeded parseBody =
+        error "handleFutureVersionParseFailure"
+
+ {-
+      undefined (unless versionOk (warning message) >> parseBody)
+        `catchParseError` \parseError -> case parseError of
+        TabsError _   -> parseFail parseError
+        _ | versionOk -> parseFail parseError
+          | otherwise -> fail message
+      where versionOk = cabalVersionNeeded <= cabalVersion
+            message   = "This package requires at least Cabal version "
+                     ++ display cabalVersionNeeded
+    -}
+
+    checkForUndefinedFlags
+        :: GenericPackageDescription
+        -> ParseResult ()
+    checkForUndefinedFlags gpd = pure ()
+{-
+        let definedFlags = map flagName flags
+        mapM_ (checkCondTreeFlags definedFlags) (maybeToList mlib)
+        mapM_ (checkCondTreeFlags definedFlags . snd) sub_libs
+        mapM_ (checkCondTreeFlags definedFlags . snd) exes
+        mapM_ (checkCondTreeFlags definedFlags . snd) tests
+
+    checkCondTreeFlags :: [FlagName] -> CondTree ConfVar c a -> PM ()
+    checkCondTreeFlags definedFlags ct = do
+        let fv = nub $ freeVars ct
+        unless (all (`elem` definedFlags) fv) $
+            fail $ "These flags are used without having been defined: "
+                ++ intercalate ", " [ n | FlagName n <- fv \\ definedFlags ]
+-}
+
+parseConfVar :: [SectionArg Position] -> ParseResult ConfVar
+parseConfVar = error "parseConfVar"
+
+-- | Parse a list of fields, given a list of field descriptions,
+--   a structure to accumulate the parsed fields, and a function
+--   that can decide what to do with fields which don't match any
+--   of the field descriptions.
+parseFields :: [FieldDescr a]      -- ^ descriptions of fields we know how to
```

Comment:
OK, well, put that in the comment, and have a reference to the function for parsing sections that *are* recursive. (Confusingly, UnrecFieldParser = unrecognized, not unrecursive!)